src/java/org/apache/cassandra/utils/memory/BufferPool.java

```diff
@@ -777,41 +817,39 @@ private void put(ByteBuffer buffer, Chunk chunk)
+
+            if (free == -1L && owner == this && owningThread == Thread.currentThread() && recycleWhenFree)
             {
-                assert owner == this;
-                // 0L => we own recycling responsibility, so must recycle;
-                // We must remove the Chunk from our local queue
+                // The chunk was fully freed, and we're the owner - let's release the chunk from this pool
+                // and give it back to the parent.
+                //
+                // We can remove the chunk from our local queue only if we're the owner of the chunk,
+                // and we're running this code on the thread that owns this local pool
+                // because the local queue is not thread safe.
+                //
+                // Please note that we may end up running `put` on a different thread when we're called
+                // from chunk.tryRecycle() on a child chunk which was previously owned by tinyPool of this pool.
+                // Such tiny chunk will point to this pool with its recycler reference. Thanks to the recycler, a thread
+                // that returns the tiny chunk can end up here in a LocalPool that's not neccessarily local to the
+                // calling thread, as there is no guarantee a child chunk is returned to the pool
+                // by the same thread that originally allocated it.
+                // It is ok we skip recycling in such case, and it does not cause
+                // a leak because those chunks are still referenced by the local pool.
                 remove(chunk);
-                chunk.recycle();
+                chunk.release();
             }
-            else if (free == -1L && owner != this && chunk.owner == null && !chunk.recycler.canRecyclePartially())
+            else if (chunk.owner == null)
             {
-                // although we try to take recycle ownership cheaply, it is not always possible to do so if the owner is racing to unset.
-                // we must also check after completely freeing if the owner has since been unset, and try to recycle
+                // The chunk has no owner, so we can attempt to recycle it from any thread because we don't need
+                // to remove it from the local pool.
+                // For normal chunk this would recycle the chunk fully or partially if not already recycled.
+                // For tiny chunk, this would recycle the tiny chunk back to the parent chunk,
+                // if this chunk is completely free.
                 chunk.tryRecycle();
             }
-            else if (chunk.owner == null && chunk.recycler.canRecyclePartially() && chunk.setInUse(Chunk.Status.EVICTED))
-            {
-                // re-cirlate partially freed normal chunk to global list
-                chunk.partiallyRecycle();
-            }
 
-            if (owner == this)
+            if (owner == this && owningThread == Thread.currentThread())
```

Comment:
As above, I would like to understand how we end up with `owner == this` but a different thread, given that `owner` is a `LocalPool`, which should only be accessed through a thread local?

Reply:
Perhaps it's due to the parent of the `LocalPool` that was introduced by the internode messaging sub-system? 

Those changes significantly increased the complexity of an already complex system. Anyway, I think your fix is good so I'm going to approve.

Reply:
> Perhaps it's due to the parent of the LocalPool that was introduced by the internode messaging sub-system?

Don't know what internode messaging subsystem has to do with all this. The issue was caused by the introduction of tiny pool with local pool as a parent.

Reply:
From what I can see, the `LocalPool tinyPool` came in with CASSANDRA-15066, `Rewrite the internode messaging subsystem`, `310a48ef`.

> Generally keeping direct references to thread-local objects in object graphs is inherently dangerous,

Definitely